### PR TITLE
Use std::num::Saturating over saturating_add_assign!()

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
There is now standard library support to make a type `Saturating` such that the arithmetic operators perform saturating math. Part of https://github.com/anza-xyz/agave/issues/502

#### Summary of Changes
Instead of using the `saturating_add_assign!()` macro, create `Saturating<T>` variables such that all operations will be saturating.

### CODE REVIEW

**Comment by steviez** [comment]
@brooksprumo - I wanted to get your thoughts on this change. The extra semantics around using `Saturating<T>` are kind of annoying (ie having to do `satuarting_var.0` to access the underlying type), but I think the benefits of having all arithmetic become saturating outweigh it.

If you like the general idea, there are a few more lines to update within `AccountsDb` crate as well as ~225 more across the mono-repo. We can do the other crates in separate PR's, but for now, I'm just looking to establish "a pattern" that we like

**Comment by brooksprumo** [comment]
> @brooksprumo - I wanted to get your thoughts on this change. The extra semantics around using `Saturating<T>` are kind of annoying (ie having to do `satuarting_var.0` to access the underlying type), but I think the benefits of having all arithmetic become saturating outweigh it.
> 
> If you like the general idea, there are a few more lines to update within `AccountsDb` crate as well as ~225 more across the mono-repo. We can do the other crates in separate PR's, but for now, I'm just looking to establish "a pattern" that we like

Seems to be OK by me. I agree that dealing with checked math properly is a bit annoying regardless of how it is addressed. I like that this impl allows general/normal math again:
```rust
stat += datapoint
```

Tagging @ilya-bobyr for general Rust thoughts too!

**Comment by illia-bobyr** [comment]
> [...] The extra semantics around using `Saturating<T>` are kind of annoying (ie having to do `satuarting_var.0` to access the underlying type), but I think the benefits of having all arithmetic become saturating outweigh it.

tl;dr: +1 to the wrapper.

---

Long story:

I worked in a codebase where we used wrapper types a lot.[1]
In my opinion, they outweigh the disadvantages of the syntactic noise that they add.

I think, it is more common to use `T::from(X)` and `.into()`, rather than directly accessing the contained value using `.0`.
But `Saturating` does not implement `From` for some reason.
There are libraries that help support this style, such as [`newtype_derive`](https://docs.rs/newtype_derive/latest/newtype_derive/) and [`derive_more`](https://crates.io/crates/derive_more).
So, it would be more familiar for me [to see many `.into()`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0d2fb38435e9a9b7d0c2ae82a2045032), rather than `Saturating(v)` (unfortunately, Rust playground does not provide `derive_more`, so you can not run the example).
`into()` is also a bit shorter :)
Notice that the `v3` and `v4` cases does not work with `Saturating`.
But with a wrapper they work just fine.

The approached we used was to avoid basic types at all.
Essentially, there would be almost no `usize`, `uX`, or `iX` types in the codebase.
Everything was named according to the semantics of the value.
This allowed to control conversions, so that only an explicitly specified conversion would work.
We also used libraries that create "vector/index" pairs.
Such that every index would only be useable with containers that it is statically typed for.
And we had wrappers that limited the range of types as well.

In my mind, the resulting code was both more readable and was better checked by the compiler.
To be fair, it was a codebase with a lot of numeric types, as it was describing a chip.
And without help from the compiler, it would be pretty easy to mix up those numeric types.
Mixing up metrics is not that big of a deal, I guess.

[1] Most folks were Haskell programmers.
In Haskell and related languages, wrapper types, called "newtype"s, is a very common idiom.
I think Haskell has a bit better support for them, but Rust is also doing a pretty good job.

**Comment by steviez** [comment]
> I think, it is more common to use T::from(X) and .into(), rather than directly accessing the contained value using .0.
But Saturating does not implement From for some reason.

Yeah, I was a bit surprised at some aspects of the implementation of `Saturating<T>`. For example, I might have expected `get()` to access underlying primitive type to match `NonZero*` types, such as this:
https://doc.rust-lang.org/nightly/std/num/type.NonZeroUsize.html#method.get

Likewise, you can do:
```rust
let mut x = std::num::Saturating(0usize);
let y = 2usize;

// Supported
x += y;
```
but not partial-eq between `Saturating<T>` and `T`:
```rust
let x = std::num::Saturating(0usize);
let y = 2usize;

// Not supported
let equal = x == y;
```

> tl;dr: +1 to the wrapper.

Appreciate the longer anecdote / justification as well, but it seems like we are in agreement that the wrapper does bring some benefit.

> Mixing up metrics is not that big of a deal, I guess.

Yeah, definitely. Also, our use of it is fairly sporadic as well; there are lots more metrics where we use non-saturating math. Some consistency / uniformity would be nice, but that is no smaller undertaking to introduce and is seemingly not a high priority thing at the moment.

**Review by illia-bobyr — DISMISSED** [review]
Not sure who would be the best person to approve it.
But just in case, LGTM.

**Comment by steviez** [comment]
I had to rebase + force push to resolve a merge conflict. I also added one more commit to remove the last several uses of the macro within the `accounts-db` crate.

Will wait for all greens on CI and then I will formally request review from ilya-bobyr & brooksprumo

File: accounts-db/src/ancient_append_vecs.rs [hunk 1]
```diff
@@ -441,13 +441,14 @@ impl AccountsDb {
...
-            rewrite_elapsed_us,
-            create_and_insert_store_elapsed_us,
-            unpackable_slots_count: 0,
-            newest_alive_packed_count: 0,
-        });
+            rewrite_elapsed_us: Saturating(rewrite_elapsed_us),
+            create_and_insert_store_elapsed_us: Saturating(create_and_insert_store_elapsed_us),
+            ..ShrinkStatsSub::default()
+        };
+        write_ancient_accounts.metrics.accumulate(&shrink_stats);
```
Comment (illia-bobyr) [inline accounts-db/src/ancient_append_vecs.rs]: *minor*

I think it is fine to use a temporary value, as the stats are not used for anything else.
And the indentation level is identical in both cases.

**Review by illia-bobyr — DISMISSED** [review]
The change looks mostly mechanical to me.
So if the CI is happy, and we agreed that this is the right direction to take in general, it should be good to go :)

File: accounts-db/src/accounts_db.rs [hunk 2]
```diff
@@ -3994,7 +3988,7 @@ impl AccountsDb {
             let (shrink_in_progress, time_us) = measure_us!(
                 self.get_store_for_shrink(slot, shrink_collect.alive_total_bytes as u64)
             );
-            stats_sub.create_and_insert_store_elapsed_us = time_us;
+            stats_sub.create_and_insert_store_elapsed_us += time_us;
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: There are many similar changes to this one, so just commenting once. Why the change from an assignment to the add-assign?

Comment (steviez) [reply accounts-db/src/accounts_db.rs]: Good call out - I probably should have included a note about this. This was actually something I learned as well; note the [following](https://doc.rust-lang.org/std/ops/):

> the assignment operator (=) has no backing trait, there is no way of overloading its semantics

The `lhs` is a `std::num::Saturating<usize>`, `rhs` is a `usize` so we can't use `=` here. Given that we just created `stats_sub` above with `ShrinkStatsSub::default()`, switching to use `+=` seemed fine from a functionality point of view. The alternative would be:
```rust
stats_sub.create_and_insert_store_elapsed_us = Saturating(time_us);
```

That being said, given that `std::num::Saturating` is `repr(transparent)`, doing the assignment is probably cheaper and godbolt agrees with me so I'll update all these:
https://godbolt.org/z/vfhPr9daT

Comment (illia-bobyr) [reply accounts-db/src/accounts_db.rs]: They should be the same.
If you turn on the optimizations, the addition is compiled to
```asm
        mov     eax, edi
        ret
```
But both are optimized away.
https://godbolt.org/z/EoToT33j1

Personally, I do not have a strong preference between `= Saturating(v)` vs `+= v`.
In this particular case, the initial value is holding all zeroes, so `+=` seems fine.

---

I am again surprised by the lack of ergonomics for `Saturating` :(

It would be so much reasonable to expect you can say `= v.into()`.
`From` and `Into` are the most used traits in Rust, I think.
Everyone knows what they mean and how to use them.

File: accounts-db/src/ancient_append_vecs.rs (hunk 1 above)
Comment (steviez) [reply accounts-db/src/ancient_append_vecs.rs]: Fair enough, not positive why I changed it 😅; done in https://github.com/anza-xyz/agave/pull/523/commits/75b44f2f9ff1aac4da079a5ab96ce5846e7cb202

File: accounts-db/src/accounts_db.rs (hunk 2 above)
Comment (steviez) [reply accounts-db/src/accounts_db.rs]: Changed in https://github.com/anza-xyz/agave/pull/523/commits/a1e70527a0c1b04ec09a0dae7b5684b596dc6848

**Review by illia-bobyr — APPROVED** [review]

File: accounts-db/src/accounts_db.rs (hunk 2 above)
Comment (steviez) [reply accounts-db/src/accounts_db.rs]: Ahh yes, good call on flipping on optimizations

> Personally, I do not have a strong preference between = Saturating(v) vs += v.

Me neither, `= Saturating(v)` more closely matches original behavior so that seems like a reasonable tiebreaker

> In this particular case, the initial value is holding all zeroes, so += seems fine.

Exactly, this was my original justification for switch to `+=` to be fine.

> I am again surprised by the lack of ergonomics for Saturating

Ya, I plan on making a PR to Rust in the next week or two when I get a chance; several lines of (simple) could seemingly improve ergonomics:
- Eq / PartialEq / PartialOrd between `T` and `Saturating<T>`
    - This will avoid having to do `.0` on the `Saturating` operand
- From / Into also seem like good candidates

**Review by brooksprumo — APPROVED** [review]
:shipit: 

lgtm, thanks!

File: accounts-db/src/accounts_db.rs (hunk 2 above)
Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs]: Thanks for the explanation; makes sense to me. The change to use
```rust
x = Saturating(123);
```
did make the review easier, which I appreciate 😸
